miner/worker.go

```diff
@@ -227,12 +231,36 @@ func (self *worker) update() {
 			self.possibleUncles[ev.Block.Hash()] = ev.Block
 			self.uncleMu.Unlock()
 		case core.TxPreEvent:
-			// Apply transaction to the pending state if we're not mining
-			if atomic.LoadInt32(&self.mining) == 0 {
-				self.currentMu.Lock()
-				self.current.commitTransactions(self.mux, types.Transactions{ev.Tx}, self.gasPrice, self.chain)
-				self.currentMu.Unlock()
+			if self.config.Automine {
+				if atomic.LoadInt32(&self.mining) == 0 {
+					self.start()
+				} else {
+					// If many txs are fired at geth at once, they will be included in blocks more quickly than this event will be processed, causing ~200+ transactions per block, and yielding many empty blocks after the burst is complete as the channel drains. This conditional simply ensure that there are actually transactions to add a next block. If not, the event is essentially ignored.
+					pending, queued := self.eth.TxPool().Stats()
+					if pending+queued == 0 {
+						continue
+
+					}
+
+					self.currentMu.Lock()
+					ww := self.current
+					self.currentMu.Unlock()
```

Comment:
Why it's not in single `Lock`? Why there is `Unlock` `noop` `Lock`?

Reply:
I had at one point thought that I might have seen a completely unquantifiable increase in performance. #nonoop